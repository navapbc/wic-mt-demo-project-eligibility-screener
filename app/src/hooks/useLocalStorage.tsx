import { SessionData } from '@customTypes/common'
import { useState } from 'react'

// Custom hook to persist state across page refresh
export default function useLocalStorage(
  key: string,
  initialValue: SessionData
) {
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = useState<SessionData>(() => {
    if (typeof window === 'undefined') {
      return initialValue
    }
    try {
      // Get from local storage by key
      const item: string | null = window.sessionStorage.getItem(key)

      // TODO: refactor type casting
      const state: SessionData = item
        ? (JSON.parse(item) as SessionData)
        : initialValue

      return state
    } catch (error) {
      // TODO: error handling
      console.log(error)
      return initialValue
    }
  })
  // Return a wrapped version of useState's setter function that persists the new value to sessionStorage.
  const setValue = (value: SessionData) => {
    try {
      // Allow value to be a function so we have same API as useState

      // TODO: refactor type casting
      const valueToStore: SessionData =
        value instanceof Function ? (value(storedValue) as SessionData) : value
      // Save state
      setStoredValue(valueToStore)
      // Save to local storage
      if (typeof window !== 'undefined') {
        window.sessionStorage.setItem(key, JSON.stringify(valueToStore))
      }
    } catch (error) {
      console.log(error)
    }
  }
  return [storedValue, setValue]
}
