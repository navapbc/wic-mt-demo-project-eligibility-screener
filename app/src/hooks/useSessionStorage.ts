import { useState } from 'react'

import { SessionData } from '@src/types'

// Custom hook to persist state across page refresh
export default function useSessionStorage(
  key: string,
  initialValue: SessionData
) {
  // Pass initial state function to useState so logic is only executed once
  const [storedValue, setStoredValue] = useState<SessionData>(() => {
    if (typeof window === 'undefined') {
      return initialValue
    }
    try {
      // Get from local storage by key
      const item: string | null = window.sessionStorage.getItem(key)

      const state: SessionData = item
        ? (JSON.parse(item) as SessionData)
        : initialValue

      return state
    } catch (error) {
      // @TODO: Production-readiness: handle more robust logging and error handling.
      console.log(error)
      return initialValue
    }
  })
  // Return a wrapped version of useState's setter function that persists the new value to sessionStorage.
  const setValue = (value: SessionData) => {
    try {
      // Allow value to be a function so we have same API as useState
      const valueToStore: SessionData =
        value instanceof Function ? (value(storedValue) as SessionData) : value
      // Save state
      setStoredValue(valueToStore)
      // Save to local storage
      if (typeof window !== 'undefined') {
        window.sessionStorage.setItem(key, JSON.stringify(valueToStore))
      }
    } catch (error) {
      console.log(error)
    }
  }
  return [storedValue, setValue]
}

export function clearSessionStorage(key: string) {
  if (typeof window === 'undefined') {
    // @TODO: Production-readiness: handle more robust logging and error handling.
    console.log('Not in client context')
  } else {
    window.sessionStorage.removeItem(key)
  }
}
