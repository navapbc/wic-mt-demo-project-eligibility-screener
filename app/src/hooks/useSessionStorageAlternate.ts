// Approach taken by Nick Scialli
// See https://typeofnan.dev/using-session-storage-in-react-with-hooks

import { useEffect, useState } from 'react'

function getSessionStorageOrDefault(key, defaultValue) {
  const stored = sessionStorage.getItem(key)
  if (!stored) {
    return defaultValue
  }
  return JSON.parse(stored)
}

export default function useSessionStorage(key, defaultValue) {
  const [value, setValue] = useState(
    getSessionStorageOrDefault(key, defaultValue)
  )

  useEffect(() => {
    sessionStorage.setItem(key, JSON.stringify(value))
  }, [key, value])

  return [value, setValue]
}
