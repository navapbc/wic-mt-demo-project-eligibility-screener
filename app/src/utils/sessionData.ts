import type {
  ChooseClinicData,
  ContactData,
  EligibilityData,
  IncomeData,
  SessionData,
} from '@src/types'

export const initialChooseClinicData: ChooseClinicData = {
  clinic: undefined,
  zipCode: '',
}

export const initialEligibilityData: EligibilityData = {
  residential: '',
  categorical: [],
  previouslyEnrolled: '',
  adjunctive: [],
}

export const initialContactData: ContactData = {
  firstName: '',
  lastName: '',
  phone: '',
  comments: '',
}

export const initialIncomeData: IncomeData = {
  householdSize: '',
}

export const initialSessionData: SessionData = {
  chooseClinic: initialChooseClinicData,
  contact: initialContactData,
  eligibility: initialEligibilityData,
  income: initialIncomeData,
}
