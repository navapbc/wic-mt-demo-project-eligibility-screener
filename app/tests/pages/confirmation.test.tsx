import { render, screen } from '@testing-library/react'
import singletonRouter from 'next/router'

import Confirmation from '@pages/confirmation'

import * as sessionModule from '@src/hooks/useSessionStorage'
import { initialSessionData } from '@utils/sessionData'

import { setMockSession, setup } from '../helpers/setup'
import { testAccessibility } from '../helpers/sharedTests'

/**
 * Test setup
 */

const route = '/confirmation'

/**
 * Begin tests
 */

it('should match full page snapshot', () => {
  const { mockSession } = setup(route)
  const { container } = render(
    <Confirmation
      sessionKey="mockSessionKey"
      session={mockSession}
      setSession={setMockSession}
    />
  )
  expect(container).toMatchSnapshot()
})

it('should pass accessibility scan', async () => {
  const { mockSession } = setup(route)
  await testAccessibility(
    <Confirmation
      sessionKey="mockSessionKey"
      session={mockSession}
      setSession={setMockSession}
    />
  )
})

it('should clear the session and redirect when start new application button is clicked', async () => {
  const { mockSession, user } = setup(route)
  // Spy on clearSessionStorage() to make sure it gets called.
  const spy = jest.spyOn(sessionModule, 'clearSessionStorage')
  render(
    <Confirmation
      sessionKey="mockSessionKey"
      session={mockSession}
      setSession={setMockSession}
    />
  )

  const button = screen.getByRole('button', { name: /Start/i })
  await user.click(button)

  expect(setMockSession).toHaveBeenCalledWith(initialSessionData)
  expect(singletonRouter).toMatchObject({ asPath: '/' })
  expect(spy).toHaveBeenCalled()
})
