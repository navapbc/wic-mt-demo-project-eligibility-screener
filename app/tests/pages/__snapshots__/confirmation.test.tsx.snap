// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match full page snapshot 1`] = `
<div>
  <h1>
    Your confirmation
  </h1>
  <p>
    Thank you for submitting your information. If you're eligible, you should hear from WIC staff at your chosen clinic within 20 days.
  </p>
  <div
    class="content-group-small"
  >
    <h2
      class="font-sans-xs"
    >
      Interested in other available assistance and benefits?
    </h2>
    <p>
      <a
        class="usa-link usa-link--external"
        href="https://dphhs.mt.gov/Assistance"
        rel="noopener noreferrer"
        target="_blank"
      >
        Learn about other assistance
      </a>
    </p>
  </div>
  <div
    class="content-group-small"
  >
    <h2
      class="font-sans-xs"
    >
      Do you need to submit another application for someone else?
    </h2>
    <button
      class="usa-button usa-button--small display-block usa-button--outline margin-top-1"
    >
      Start a new application
    </button>
  </div>
  <div
    class="content-group-small"
  >
    <h2
      class="font-sans-xs"
    >
      Keep a copy of your confirmation for your records.
    </h2>
  </div>
  <div
    class="review-collection margin-top-6 border-bottom-1px"
  >
    <h2>
      Eligibility questions
    </h2>
    <dl
      class="margin-bottom-2"
    >
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Do you live or work in Montana?
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        >
          Eligibility.
        </dd>
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Please select all the following that apply to your household:
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        >
          <ul />
        </dd>
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Have you or someone in your household received WIC benefits before?
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        >
          Eligibility.
        </dd>
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Are you or someone in your household currently enrolled in any of the following programs in Montana?
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        >
          <ul />
        </dd>
      </div>
    </dl>
  </div>
  <div
    class="review-collection margin-top-6 border-bottom-1px"
  >
    <h2>
      Choose a clinic
    </h2>
    <dl
      class="margin-bottom-2"
    >
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Residential ZIP code
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        >
          <div />
        </dd>
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            You selected this WIC Clinic to visit
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        />
      </div>
    </dl>
  </div>
  <div
    class="review-collection margin-top-6 border-bottom-1px"
  >
    <h2>
      Contact information
    </h2>
    <dl
      class="margin-bottom-2"
    >
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            First name
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        />
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Last name
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        />
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Phone number
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        />
      </div>
      <div
        class="review-element margin-bottom-3"
      >
        <dt>
          <strong>
            Comments, questions, other information
          </strong>
        </dt>
        <dd
          class="margin-left-0"
        />
      </div>
    </dl>
  </div>
</div>
`;
